src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    internal class TdsWriter
+    {
+        private readonly TdsWriteStream _writeStream;
+
+        // Header length constants
+        private const int OutputHeaderLen = TdsEnums.HEADER_LEN;
+
+        // Out buffer variables
+        private readonly byte[] _outBuff; // internal write buffer - initialize on login
```

Comment:
Although I don't think the current implementation is necessary (as per Saurabh's comments), I do think it's a good idea to have a preinitialized buffer for storing bytes before sending them all to the stream. This is one benefit to this approach over extension methods - we don't have to share a buffer between all the writers.

However, as it's currently written, there's no thread safety within the writer. Concurrent requests to write a number could trample over the buffer. If this is expected/allowed, then no action necessary. But I think we should at least consciously make that decision.

Reply:
We don't expect concurrent requests on 1 stream, so allocation tampering is not of concern. However, we will review this as design progresses. At the moment, our goal is to limit memory usage by reusing buffers.